src/_P044_P1WifiGateway.ino

```diff
@@ -5,127 +5,223 @@
+ 
 
+struct P044_Task : public PluginTaskData_base {
 
-struct P044_data_struct : public PluginTaskData_base {
+  enum ParserState : byte {
+    P044_WAITING,
+    P044_READING,
+    P044_CHECKSUM
+  };
 
-  P044_data_struct() {
+  P044_Task() {
     clearBuffer();
   }
 
-  ~P044_data_struct() {
+  ~P044_Task() {
     stopServer();
   }
 
+  inline static bool serverActive(WiFiServer *server) {
+#if defined(ESP8266)
+    return nullptr != server && server->status() != CLOSED;
+#elif defined(ESP32)
+    return nullptr != server && *server;
+#endif
+  }
+
   void startServer(unsigned int portnumber) {
+    if (gatewayPort == portnumber && serverActive(P1GatewayServer)) {
+      // server is already listening on this port
+      return;
+    }
     stopServer();
+    gatewayPort = portnumber;
     P1GatewayServer = new WiFiServer(portnumber);
-    if (nullptr != P1GatewayServer) {
+    if (nullptr != P1GatewayServer && WiFi.isConnected()) {
       P1GatewayServer->begin();
-      addLog(LOG_LEVEL_DEBUG, String(F("P1   : WiFi server started at port ")) + portnumber);
+      if(serverActive(P1GatewayServer)) {
+        addLog(LOG_LEVEL_INFO, String(F("P1   : WiFi server started at port ")) + portnumber);
+      } else {
+        addLog(LOG_LEVEL_ERROR, String(F("P1   : WiFi server start failed at port ")) +
+            portnumber + String(F(", retrying...")));
+      }
+    }
+  }
+
+  void checkServer() {
+    if (!serverActive(P1GatewayServer) && WiFi.isConnected()) {
+      P1GatewayServer->close();
+      P1GatewayServer->begin();
+      if(serverActive(P1GatewayServer)) {
+        addLog(LOG_LEVEL_INFO, F("P1   : WiFi server started"));
+      }
     }
   }
 
   void stopServer() {
-    clearBuffer();
     if (nullptr != P1GatewayServer) {
+      if (P1GatewayClient) P1GatewayClient.stop();
       P1GatewayServer->close();
-      addLog(LOG_LEVEL_DEBUG, F("P1   : WiFi server closed"));
+      addLog(LOG_LEVEL_INFO, F("P1   : WiFi server closed"));
       delete P1GatewayServer;
       P1GatewayServer = nullptr;
     }
   }
 
-  void clearBuffer() {
-    serial_buffer = "";
-    serial_buffer.reserve(P044_BUFFER_SIZE);
-    bytes_read = 0;
+  bool hasClientConnected() {
+    if (P1GatewayServer->hasClient())
+    {
+      if (P1GatewayClient) P1GatewayClient.stop();
+      P1GatewayClient = P1GatewayServer->available();
+      P1GatewayClient.setTimeout(CONTROLLER_CLIENTTIMEOUT_DFLT);
+      addLog(LOG_LEVEL_INFO, F("P1   : Client connected!"));
+    }
+
+    if (P1GatewayClient.connected())
+    {
+      clientConnected = true;
+    }
+    else
+    {
+      if (clientConnected) // there was a client connected before...
+      {
+        clientConnected = false;
+        addLog(LOG_LEVEL_INFO, F("P1   : Client disconnected!"));
+      }
+    }
+    return clientConnected;
   }
 
-  bool isInit() const {
-  	return nullptr != P1GatewayServer;
+  void handleClientIn() {
+    uint8_t net_buf[P044_NETBUF_SIZE];
```

Comment:
`uint8_t net_buf[P044_NETBUF_SIZE] = {0};`
So it will be initialized with 0 values.
You may still want to 0-terminate it after writing, but at least the received string will also be null-terminated if `net_bytes_read < count`

Reply:
Just moved this code without altering it. See remark in ticket about this part of code being superfluous and imo can be removed.

Reply:
Ah if it can be removed, please do :)

Reply:
Referring explicitly to my question:
> The file header of P1WifiGateway states that the gateway code is based on P020 Ser2Net.
> Are the "Reset target after boot" and "data forwarding from the WifiClient to the serial-out" left-overs of Ser2Net (as the P1Meter does not receive data)?
> If so, both parts imo can be removed and the 'now hidden' status led pin can be shown in the gui and made configurable instead.

Do you know the answer (as my opinion is just what it is: an opinion)?
Especially, the second one is somewhat more tricky as it also impacts the plugin configuration (users needs to reconfigure after upgrade)

Reply:
> Are the "Reset target after boot" and "data forwarding from the WifiClient to the serial-out" left-overs of Ser2Net (as the P1Meter does not receive data)?

I don't know.
The P1 meter is indeed "send only" so there is no need at all for such a communication back.
Only thing you should watch out for is to change the meaning of existing settings out there.
The plugin has been used for quite a while, so there are existing settings out there for this plugin.
In very specific cases we can make a settings transition, based on the "BUILD" value, which is also stored in the settings. But that's only a forward change, so it will then alter the settings and loading an older build again will then use the altered settings.

Reply:
Ok, removed client forwarding data to serial